File: src/980-unique-paths-iii/Solution.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

public class Solution {
    public int UniquePathsIII(int[][] grid) {
        int empty = 0, sx = 0, sy = 0;
        
        for (int r = 0; r < grid.Length; r++) {
            for (int c = 0; c < grid[0].Length; c++) {
                if (grid[r][c] == 0) {
                    empty++;
                } else if (grid[r][c] == 1) {
                    sx = r;
                    sy = c;
                }
            }
        }
        return dfs(grid, sx, sy, empty);
    }
    
    private int dfs(int[][] grid, int sx, int sy, int empty) {
        if (sx < 0 || sy < 0 || sx >= grid.Length || 
            sy >= grid[0].Length || grid[sx][sy] == -1) {
            return 0;
        }
        
        if (grid[sx][sy] == 2) {
            return empty < 0 ? 1 : 0;
        }
        grid[sx][sy] = -1;
        empty -= 1;
        
        int paths = dfs(grid, sx + 1, sy, empty);
        paths += dfs(grid, sx, sy + 1, empty);
        paths += dfs(grid, sx - 1, sy, empty);
        paths += dfs(grid, sx, sy - 1, empty);
        
        grid[sx][sy] = 0;
        empty += 1;
        
        return paths;
    }
}

public class Driver
{
    public static void Main(String[] args)
    {
        Solution solution = new Solution();
        StreamReader inputfile = new StreamReader("input.txt");
        StreamReader outputfile = new StreamReader("output.txt");
        String line;
        int testno = 1;
        while ((line = inputfile.ReadLine()) != null)
        {
            int[][] grid = StringToInt2dArray(line);
            
            String expected = outputfile.ReadLine();
            int result = solution.UniquePathsIII(grid);
            String output = result.ToString();
            CheckAnswer(expected, output, testno++);
        }
    }
    public static void CheckAnswer(String expected, String output, int testno)
    {
        if (expected.Equals(output))
        {
            Console.WriteLine("TEST {0} SUCCESS EXPECTED: {1} OUTPUT: {2}", testno, expected, output);
        }
        else
        {
            Console.WriteLine("TEST {0} FAILED EXPECTED: {1} OUTPUT: {2}", testno, expected, output);
        }
    }

    public static bool StringToBoolean(String input) {
        return input.Equals("true") ? true : false;
    }

    public static String BooleanToString(bool input) {
        return input ? "true" : "false";
    }

    public static int StringToInteger(String input) {
        return int.Parse(input);
    }

    public static String IntegerArrayToString(int[] nums)
    {
        int length = nums.Length;
        if (length == 0)
        {
            return "[]";
        }

        String result = "";
        for (int index = 0; index < length; index++)
        {
            int number = nums[index];
            result += number.ToString() + ",";
        }
        return "[" + result.Substring(0, result.Length - 1) + "]";
    }

    public static int[] StringToIntegerArray(String input)
    {
        input = input.Trim();
        input = input.Substring(1, input.Length - 2);
        if (input.Length == 0)
        {
            return new int[0];
        }

        String[] parts = input.Split(',');
        int[] output = new int[parts.Length];
        for (int index = 0; index < parts.Length; index++)
        {
            String part = parts[index].Trim();
            output[index] = int.Parse(part);
        }
        return output;
    }

    public static int[][] StringToInt2dArray(String input)
    {
        input = input.Trim();
        input = input.Substring(1, input.Length - 2);
        if (input.Length == 0)
        {
            return new int[0][];
        }
        List<int[]> list = new List<int[]>();
        int startIndex = 0, endIndex = 0;
        while (true)
        {
            startIndex = input.IndexOf(
                '[', startIndex,
                input.Length - startIndex);

            if (startIndex < 0)
                break;

            endIndex = input.IndexOf(
                ']', startIndex,
                input.Length - startIndex);

            String[] parts = input.Substring(startIndex + 1, endIndex - startIndex - 1).Split(',');
            int[] arr = new int[parts.Length];
            for (int i = 0; i < arr.Length; i++)
            {
                arr[i] = int.Parse(parts[i].Trim());
            }
            list.Add(arr);
            startIndex = endIndex + 1;
        }

        return list.ToArray();
    }

    public static List<int> StringToIntegerArrayList(String input) {
        input = input.Trim();
        input = input.Substring(1, input.Length - 2);
        String[] parts = input.Split(',');
        var output = new List<int>();
        for(int index = 0; index < parts.Length; index++) {
            String part = parts[index].Trim();
            output.Add(int.Parse(part));
        }
        return output;
    }

    public static String IntegerArrayListToString(IList<int> nums, int length) {
        if (length == 0) {
            return "[]";
        }
    
        String result = "";
        for(int index = 0; index < length; index++) {
            int number = nums[index];
            result += number + ", ";
        }
        return "[" + result.Substring(0, result.Length - 2) + "]";
    }
    
    public static String IntegerArrayListToString(IList<int> nums) {
        return IntegerArrayListToString(nums, nums.Count);
    }

    public static String Int2dArrayToString(int[][] array) {
        if (array == null) {
            return "null";
        }
        if (array.Length == 0) {
            return "[]";
        }
    
        StringBuilder sb = new StringBuilder("[");
        foreach (int[] item in array) {

            sb.Append(IntegerArrayToString(item));
            sb.Append(",");
        }
    
        sb[sb.Length - 1] = ']';
        return sb.ToString();
    }
}

File: src/980-unique-paths-iii/Solution.vb
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Text

Class Solution 
	Public Function UniquePathsIII(grid As Integer()()) As Integer
		Dim empty = 0
		Dim sx = 0
		Dim sy = 0

		For r As Integer = 0 To grid.Length - 1
			For c As Integer = 0 To grid(0).Length - 1
				If grid(r)(c) = 0 Then
					empty += 1
				ElseIf grid(r)(c) = 1 Then
					sx = r
					sy = c
				End If
			Next
		Next
		Return Dfs(grid, sx, sy, empty)
	End Function

	Private Function Dfs(grid As Integer()(), sx As Integer, sy As Integer, empty As Integer) As Integer
		If sx < 0 OrElse sy < 0 OrElse sx >= grid.Length OrElse sy >= grid(0).Length OrElse grid(sx)(sy) = -1 Then
			Return 0
		End If
		If grid(sx)(sy) = 2 Then
			Return IIf(empty < 0, 1, 0)
		End If
		grid(sx)(sy) = -1
		empty -= 1

		Dim paths = Dfs(grid, sx + 1, sy, empty)
		paths += Dfs(grid, sx, sy + 1, empty)
		paths += Dfs(grid, sx - 1, sy, empty)
		paths += Dfs(grid, sx, sy - 1, empty)

		grid(sx)(sy) = 0
		empty += 1

		Return paths
	End Function
	
End Class

Public Module Driver
	Public Sub Main()
		Dim inputfile As StreamReader = new StreamReader("input.txt")
		Dim outputfile As StreamReader = new StreamReader("output.txt")

		Dim solution = new Solution()
		Dim testno = 1
		While True 
			Dim line As String = inputfile.ReadLine()
			If IsNothing(line) Then
				Exit While
			End If
			Dim grid = StringToInt2dArray(line)

			Dim expected = outputfile.ReadLine()
			Dim result = solution.UniquePathsIII(grid)
			Dim output = result.ToString()
			CheckAnswer(expected, output, testno)
			testno += 1
		End While
	End Sub

	Public Sub CheckAnswer(expected As String, output As String, testno As Integer)
		If expected.Equals(output) Then
			Console.WriteLine("TEST {0} SUCCESS EXPECTED: {1} OUTPUT: {2}", testno, expected, output)
		Else
			Console.WriteLine("TEST {0} FAILED EXPECTED: {1} OUTPUT: {2}", testno, expected, output)
		End If
	End Sub

	Public Function StringToInteger(input As String) As Integer
		Return Integer.Parse(input)
	End Function

	Public Function StringToBoolean(input As String) As Boolean
		Return IIf(input.Equals("true"), True, False)
	End Function

	Public Function BooleanToString(input As Boolean) As String
		Return IIf(input, "true", "false")
	End Function

	Public Function IntegerArrayToString(nums As Integer()) As String
		Dim length As Integer = nums.Length
		If length = 0 Then
			Return "[]"
		End If
		Dim result As String = ""
		For index = 0 To length - 1
			Dim number As Integer = nums(index)
			result += number.ToString() + ","
		Next
		Return "[" + result.Substring(0, result.Length() - 1) + "]"
	End Function

	Public Function StringToIntegerArray(input As String) As Integer() 
		input = input.Trim()
		input = input.Substring(1, input.Length - 2)
		If input.Length = 0 Then
			return New Integer(0) {}
		End If
		Dim parts As String() = input.split(","c)
		Dim output As Integer() = New Integer(parts.Length - 1) {}

		For index = 0 To parts.Length - 1
			Dim part As String = parts(index).Trim()
			output(index) = Integer.Parse(part)
		Next
		Return output
	End Function

	Public Function StringToInt2dArray(input As String) As Integer()()
		input = input.Trim()
		input = input.Substring(1, input.Length - 2)
		If input.Length = 0 Then
			return New Integer(0)() {}
		End If
		Dim list as List(Of Integer()) = new List(Of Integer())
		Dim startIndex As Integer = 0
		Dim endIndex As Integer = 0

		While True
			startIndex = input.IndexOf("["c, startIndex, input.Length - startIndex)

			If startIndex < 0
				Exit While
			End If

			endIndex = input.IndexOf("]"c, startIndex, input.Length - startIndex)

			Dim parts As String() = input.Substring(startIndex + 1, endIndex - startIndex - 1).Split(","c)
			Dim arr As Integer() = new Integer(parts.Length - 1) {}
			For i = 0 to arr.Length - 1
				arr(i) = Integer.Parse(parts(i).Trim())
			Next
			list.Add(arr)
			startIndex = endIndex + 1
		End While
		Return list.ToArray()
	End Function

	Public Function StringToIntegerArrayList(input As String) As List(Of Integer)
		input = input.Trim()
		input = input.Substring(1, input.Length - 2)
		Dim parts As String() = input.SPlit(","c)
		Dim output = New List(Of Integer)()
		For index = 0 To parts.Length - 1
			Dim part = parts(index).Trim()
			output.Add(Integer.Parse(part))
		Next
		Return output
	End Function

	Public Function IntegerArrayListToString(nums As List(Of Integer), length As Integer) As String
		If length = 0 Then
			Return "[]"
		End If
		Dim result = ""
		For index = 0 To length - 1
			Dim number = nums(index)
			result += number.ToString() + ", "
		Next
		Return "[" + result.Substring(0, result.Length - 2) + "]"
	End Function

	Public Function IntegerArrayListToString(nums As List(Of Integer)) As String
		Return IntegerArrayListToString(nums, nums.Count)
	End Function
	
	Public Function Int2dArrayToString(array As Integer()()) As String
		If array Is Nothing 
			Return "null"
		End If
		If array.Length = 0 Then
			Return "[]"
		End If
		Dim sb = New StringBuilder("[")
		For Each item As Integer() In array
			sb.Append(IntegerArrayToString(item))
			sb.Append(",")
		Next
		sb(sb.Length - 1) = "]"c
		Return sb.ToString()
	End Function
End Module